### DISCUSSION TITLE
Full precision of nanosecond timestamps is not preserved

### DISCUSSION THREAD

**Discussion by philrz** [zeek/zeek#1163] [Projects]
Consider the attached small pcapng.

[pings.pcapng.zip](https://github.com/zeek/zeek/files/5209462/pings.pcapng.zip)

When viewed in Wireshark, the timestamps have granularity out to 9 digits after the decimal place.

![image](https://user-images.githubusercontent.com/5934157/92934553-79438e00-f3fc-11ea-91c5-4b2ee9e447d0.png)

However, when run through Zeek, only 6 digits are preserved.

```
$ /usr/local/zeek-3.2.1/bin/zeek -C -r pings.pcapng local

$ cat conn.log | /usr/local/zeek-3.2.1/bin/zeek-cut ts
1599787799.006763
1599787801.385053
1599787804.008506
1599787806.320433
```

I [asked about this on Slack](https://zeekorg.slack.com/archives/CSZBXF6TH/p1599784243098800) and @grigorescu confirmed by pointing to https://github.com/zeek/zeek/blob/master/src/threading/Formatter.cc#L113 that 6 is indeed the current limit. Vern also joined the thread:

> **Vern Paxson**: Historically, libpcap provided at best usec precision, hence the wiring-in of 6 digits.  However, this has been improved, so Zeek should allow more precision.  OTOH, it’s a pity to bloat logs (slightly!) with digits that are almost never of any interest

> **Phil Rzewski**: Perhaps room for a configurable option that's off by default, so folks only pay for the bloat in the logs if it really matters to them? Like @timwoj  mentioned, I had a heck of a time even finding pcapng samples online that went out to that level of precision, and I had to invoke non-default tcpdump options on a system with a particular NIC to even generate my own. But I happened to be working with a user that did have legit pcapng files that went out that far, which was how I bumped into this.

> **Vern Paxson**: Yeah, I like that notion.

As noted, I had a difficult time finding sample data online, so the synthetic sample linked above was taken on an AWS Linux host via `tcpdump -i ens4 -w pings.pcap --time-stamp-precision=nano net 192.168.1.0/24` while I ran pings in another window.

**Comment by timwoj** [comment]
Some comments on an earlier PR that attempted to fix the issue described here, but fell short of what was actually needed (https://github.com/zeek/zeek/pull/1171):

From @jsiwek:

> - Not that I'm sure the new formatting option should effect JSON output, but wonder about the reason for JSON output generally having less precision (maybe it's actually related to the 2 points below):
> 
> ```
> $ zeek -r pings.pcapng LogAscii::use_json=T Log::timestamp_precision=9
> $ jq '.ts' conn.log | head -n1
> $ 1599787799.006763
> ```
> 
> 
> - Agree with @duffy-corelight that it's important to improve the underlying data type for times/intervals since otherwise these higher precision timestamps won't get meaningfully distinct representations using `double`. (Thinking that does need to happen before or along with this PR).
> 
> - Think we have to do further changes so that we're truly reading nanosecond-timestamps from the pcap: we're taking the timestamp from libpcap packet header using `timeval` to mean "second and microsecond". E.g. here's a quick googling of how to ask libpcap to do that I only looked at superficially: https://stackoverflow.com/a/17637187
> 

From @duffy-ocraven :

> Upon examining https://github.com/zeek/spicy/blob/master/hilti/include/rt/types/interval.h, I believe that has implemented the best solution for preserving well-behaved comparison operators, and is one which Zeek could adopt if the enhancement to timestamp_precision beyond 6 decimals after the period goes forward. In https://github.com/zeek/spicy/blob/master/hilti/src/rt/types/time.cc is the key paradigm:
> 
> ```
> // NOLINTNEXTLINE(bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions)
>     double frac = (_nsecs.Ref() % 1000000000) / 1e9;
> ```
> 
> which only assigns a double after throwing away all the whole seconds (planning to pick them up again as needed) from where they are in a
> 
> ```
> private:
>     integer::safe<int64_t> _nsecs
> ```
> 
> for interval, or a
> 
> ```
> private:
>     integer::safe<uint64_t> _nsecs
> ```
> 
> for time.
> 
> Also in that latter file is where Time::currentTime() is implemented, which it does in terms of gettimeofday() and using the timeval.tv_usec which is microseconds.

**Comment by philrz** [comment]
I bumped into what looks like a variation on what's described here. I'm logging it here to make it easier for me to refer back to as a known issue. A sample pcap that reproduces the issue is https://archive.wrccdc.org/pcaps/2018/wrccdc.2018-03-24.101533000000000.pcap.gz. I'm using Zeek v4.0.0 on macOS with the only modification being the addition of the https://github.com/corelight/json-streaming-logs plugin.

Generating the logs:

```
$ /usr/local/zeek-4.0.0/bin/zeek -r wrccdc.2018-03-24.101533000000000.pcap "JSONStreaming::enable_log_rotation=F" local
WARNING: No Site::local_nets have been defined.  It's usually a good idea to define your local networks.
```

The issue shows up in one of the `x509` events:

```
$ cat x509.log | /usr/local/zeek/bin/zeek-cut ts id certificate.not_valid_after | grep 2.385642e+09
1521911732.519299	Fjasql3YGDgnZF8d73	2.385642e+09
```

Grabbing the equivalent event in its JSON representation:

```
$ cat json_streaming_x509.log | jq '. | select (.id=="Fjasql3YGDgnZF8d73")'
{
  "_path": "x509",
  "_write_ts": "2018-03-24T17:15:32.519299Z",
  "ts": "2018-03-24T17:15:32.519299Z",
  "id": "Fjasql3YGDgnZF8d73",
  "certificate.version": 3,
  "certificate.serial": "C5F8CDF3FFCBBF2D",
  "certificate.subject": "CN=10.150.0.85,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU",
  "certificate.issuer": "CN=10.150.0.85,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU",
  "certificate.not_valid_before": "2018-03-22T14:22:37.000000Z",
  "certificate.not_valid_after": "2045-08-06T14:22:37.000000Z",
  "certificate.key_alg": "rsaEncryption",
  "certificate.sig_alg": "sha256WithRSAEncryption",
  "certificate.key_type": "rsa",
  "certificate.key_length": 2048,
  "certificate.exponent": "65537",
  "basic_constraints.ca": true
}
```

If we expand the `2.385642e+09` from the TSV log to `2385642000` and plug it into a [timestamp converter](https://www.timestamp-converter.com/), we see its ISO representation is `2045-08-06T14:20:00.000Z`, which is off by 2+ minutes compared to the one in the `2045-08-06T14:22:37.000000Z`. I'm guessing this may be related to the time being out beyond the end of epoch time, but I don't know the Zeek code, so that's just speculation on my part.

If this is considered different enough to justify its own separate issue (i.e. this seems to be more about accuracy than just precision), I'd be happy to open one.

**Reply by duffy-ocraven** [reply]
Phil,
Welcome to the wild-'n-wooly world of time-date-programming. A delicious programming bug, but indeed different enough to justify its own separate issue. The
  "certificate.not_valid_after": "2045-08-06T14:22:37.000000Z"
processing code shouldn't try to "capture" what can fit in just seven decimal digits, when that e+09 part clearly shows that there are e+09 possible different values in the range from 0.00000000e+09 through 9.99999999e+09.
 
It appears the x509.log generating code has a string-renderer as %.7g or %.7e or something
>>> "%.6g" % (2.385642157e+09,)
'2.38564e+09'
>>> "%.7g" % (2.385642157e+09,)
'2.385642e+09'
>>> "%.9g" % (2.385642157e+09,)
'2.38564216e+09'
>>> "%.10g" % (2.385642157e+09,)
'2385642157'

which means that it simply says "heck, toss away the final two or three digits of meaningful distinctions. In the case of 2.385642e+09 'Sun Aug  6 10:20:00 2045' vs 2.385642157e+09 that is the 157 seconds.

>>> import time
>>> time.ctime(2.385642e+09)
'Sun Aug  6 10:20:00 2045'
>>> time.ctime(2.385642157e+09)
'Sun Aug  6 10:22:37 2045'
>>> time.strptime( "2045-08-06T14:22:37.000000Z", '%Y-%m-%dT%H:%M:%S.000000Z')
time.struct_time(tm_year=2045, tm_mon=8, tm_mday=6, tm_hour=14, tm_min=22, tm_sec=37, tm_wday=6, tm_yday=218, tm_isdst=-1)

**Reply by philrz** [reply]
Thanks @duffy-ocraven! Per your recommendation, I opened a separate issue https://github.com/zeek/zeek/issues/1450 for that one.
